doc/knn.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "KNN" "1" "September 2017" "www.complex-networks.net" "www.complex-networks.net"
.
.SH "NAME"
\fBknn\fR \- Compute the average nearest neighbours degree function
.
.SH "SYNOPSIS"
\fBknn\fR \fIgraph_in\fR [\fINO|LIN|EXP\fR \fIbin_param\fR]
.
.SH "DESCRIPTION"
\fBknn\fR computes the average nearest neighbours degree function knn(k) of the graph \fIgraph_in\fR given as input\. The program can (optionally) average the results over bins of equal or exponentially increasing width (the latter is also known as logarithmic binning)\.
.
.SH "PARAMETERS"
.
.TP
\fIgraph_in\fR
undirected input graph (edge list)\. If is equal to \fB\-\fR (dash), read the edge list from STDIN\.
.
.TP
NO
If the second (optional) parameter is equal to \fBNO\fR, or omitted, the program will print on output the values of knn(k) for all the degrees in \fIgraph_in\fR\.
.
.TP
LIN
If the second (optional) parameter is equal to \fBLIN\fR, the program will average the values of knn(k) over \fIbin_param\fR bins of equal length\.
.
.TP
EXP
If the second (optional) parameter is equal to \fBEXP\fR, the progam will average the values of knn(k) over bins of exponentially increasing width (also known as \'logarithmic binning\', which is odd, since the width of subsequent bins increases exponentially, not logarithmically, but there you go\.\.\.)\. In this case, \fIbin_param\fR is the exponent of the increase\.
.
.TP
\fIbin_param\fR
If the second parameter is equal to \fBLIN\fR, \fIbin_param\fR is the number of bins used in the linear binning\. If the second parameter is \fBEXP\fR, \fIbin_param\fR is the exponent used to determine the width of each bin\.
.
.SH "OUTPUT"
The output is in the form:
.
.IP "" 4
.
.nf

    k1 knn(k1)
    k2 knn(k2)
    \.\.\.\.
.
.fi
.
.IP "" 0
.
.P
If no binning is selected, \fBk1\fR, \fBk2\fR, etc\. are the degrees observed in \fIgraph_in\fR\. If linear or exponential binning is required, then \fBk1\fR, \fBk2\fR, etc\. are the right extremes of the corresponding bin\.
.
.SH "EXAMPLES"
To compute the average neanest\-neighbours degree function for a given graph we just run:
.
.IP "" 4
.
.nf

      $ knn er_1000_5000\.net
      2 10\.5
      3 11\.333333
      4 10\.785714
      5 11\.255319
      6 11\.336601
      7 11\.176292
      8 11\.067568
      9 11\.093519
      10 10\.898438
      11 10\.906009
      12 11\.031353
      13 10\.73938
      14 10\.961538
      15 10\.730864
      16 10\.669118
      17 10\.702206
      18 10\.527778
      19 11\.302632
      20 11\.8
      $
.
.fi
.
.IP "" 0
.
.P
Since we have not requested a binning, the program will output the value of knn(k) for each of the degrees actually observed in the graph \fBer_1000_5000\.net\fR (the mininum degree is 2 and the maximum degree is 20)\. Notice that in this case, as expected in a graph without degree\-degree correlations, the values of knn(k) are almost independent of k\.
.
.P
We can also ask \fBknn\fR to bin the results over 5 bins of equal width by running:
.
.IP "" 4
.
.nf

    $ knn er_1000_5000\.net LIN 5
    6 11\.249206
    10 11\.037634
    14 10\.919366
    18 10\.68685
    22 11\.474138
    $
.
.fi
.
.IP "" 0
.
.P
Let us consider the case of \fBmovie_actors\.net\fR, i\.e\. the actors collaboration network\. Here we ask \fBknn\fR to compute the average nearest\-neighbours degrees using exponential binning:
.
.IP "" 4
.
.nf

    $ knn movie_actors\.net EXP 1\.4
    2 142\.56552
    5 129\.09559
    9 158\.44493
    15 198\.77922
    23 205\.96538
    34 210\.07379
    50 227\.57167
    72 235\.89857
    102 254\.47583
    144 276\.572
    202 307\.11004
    283 337\.83733
    397 370\.34222
    556 410\.89117
    779 446\.66331
    1091 498\.73118
    1527 547\.31923
    2137 577\.87852
    2991 582\.6855
    4187 557\.44801
    $
.
.fi
.
.IP "" 0
.
.P
Notice that, due to the presence of the second parameter \fBEXP\fR, the program has printed on output knn(k) over bins of exponentially increasing width, using an exponent \fB1\.4\fR\. This is useful for plotting with log or semilog axes\. In this case, the clear increasing trend of knn(k) indicates the presence of assortative correlations\.
.
.SH "SEE ALSO"
knn_w(1), deg_seq(1)
.
.SH "REFERENCES"
.
.IP "\(bu" 4
V\. Latora, V\. Nicosia, G\. Russo, "Complex Networks: Principles, Methods and Applications", Chapter 7, Cambridge University Press (2017)
.
.IP "" 0
.
.SH "AUTHORS"
(c) Vincenzo \'KatolaZ\' Nicosia 2009\-2017 \fB<v\.nicosia@qmul\.ac\.uk>\fR\.


debian/changelog
netbunch (1.0-1) unstable; urgency=medium

  * Initial release 

 -- Vincenzo (KatolaZ) Nicosia <katolaz@freaknet.org>  Thu, 28 Sep 2017 06:46:47 +0100


src/knn/knn.c
/**
 *  This program is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 *  (c) Vincenzo Nicosia 2009-2017 -- <v.nicosia@qmul.ac.uk>
 * 
 *  This file is part of NetBunch, a package for complex network
 *  analysis and modelling. For more information please visit:
 *
 *             http://www.complex-networks.net/
 *
 *  If you use this software, please add a reference to 
 *
 *               V. Latora, V. Nicosia, G. Russo             
 *   "Complex Networks: Principles, Methods and Applications"
 *              Cambridge University Press (2017) 
 *                ISBN: 9781107103184
 *
 ***********************************************************************
 *
 * Compute knn(k) for a given unweighted graph. If requested, the
 * average is computed for ranges of degrees falling into a number of
 * equal bins or into bins or exponentially increasing width.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>

#include "utils.h"
#include "knntree.h"

#define NO_BIN   0x01
#define BIN_LIN  0x02
#define BIN_EXP  0x04

void usage(char *argv[]){

  printf("********************************************************************\n"
         "**                                                                **\n"
         "**                        -*-     knn    -*-                      **\n"
         "**                                                                **\n"
         "**   Compute the average nearest-neighbour degree function knn(k) **\n"
         "**   of the unweighted graph provided as input.                   **\n"
         "**                                                                **\n"
         "**   The input file 'graph_in' is an edge-list:                   **\n"
         "**                                                                **\n"
         "**                            I_1 J_1                             **\n"
         "**                            I_2 J_2                             **\n"
         "**                            I_3 J_3                             **\n"
         "**                            ... ...                             **\n"
         "**                            I_K J_K                             **\n"
         "**                                                                **\n"
         "**    If 'graph_in' is equal to '-' (dash), read the file from    **\n" 
         "**    the standard input (STDIN).                                 **\n"
         "**                                                                **\n"
         "**    If no second parameter is given (or the second parameter    **\n"
         "**    is equal to 'NO'), the program prints on output the value   **\n"
         "**    of knn(k) for each degree in the graph, in the format:      **\n"
         "**                                                                **\n"
         "**          k1  knn(k1)                                           **\n"
         "**          k2  knn(k2)                                           **\n"
         "**          .....                                                 **\n"
         "**                                                                **\n"
         "**     If the second parameter is 'LIN', the values of knn(k)     **\n"
         "**     will be grouped into 'bin_param' equal bins.               **\n"
         "**                                                                **\n"
         "**     If the second parameter is 'EXP;, the values of knn(k)     **\n"
         "**     will be grouped into bins whose width increases            **\n"
         "**     exponentially with exponent 'bin_param' (also known as     **\n"
         "**     'logarithmic binning').                                    **\n"
         "**                                                                **\n"
         "********************************************************************\n"
         " This is Free Software - You can use and distribute it under \n"
         " the terms of the GNU General Public License, version 3 or later\n\n"
         " Please visit http://www.complex-networks.net for more information\n\n"
         " (c) Vincenzo Nicosia 2010-2017 (v.nicosia@qmul.ac.uk)\n"
         "********************************************************************\n\n"
         );
  printf("Usage: %s <graph_in> [<NO|LIN|EXP> <bin_param>]\n", argv[0]);
}



typedef struct{
  unsigned int num_bins;
  unsigned int *degs;
  unsigned int *num;
  double *values;
} knn_hist_t;


/* compute the sum of the degrees of the neighbours of the node with
   label 'idx' */
double neigh_degs_sum(unsigned int *J_slap, unsigned int *r_slap, unsigned int idx){

  double sum = 0;
  unsigned j, k;
  
  for(j=r_slap[idx]; j < r_slap[idx+1]; j++){
    k = J_slap[j];
    sum += r_slap[k+1] - r_slap[k];
  }
  return sum;
}

/* compute knn(k) for each degree k */
void compute_knn(unsigned int *J_slap, unsigned int *r_slap, unsigned int N, 
                 knntree_t t){

  knnsum_t elem, *ptr;
  unsigned int i;

  for(i=0; i<N; i++){
    elem.k = r_slap[i+1] - r_slap[i];
    elem.knnsum = neigh_degs_sum(J_slap, r_slap, i);
    ptr = iltree_lookup(t, &elem);
    if (!ptr){
      elem.Nk = 1;
      iltree_insert(t, &elem);
    }
    else{
      ptr->Nk +=1;
      ptr->knnsum += elem.knnsum;
    }
  }
}

/* aggregate all the values of knn(k) for the degrees falling in a
   certain bin */
void aggregate_knn(void *elem, void *h){

  knn_hist_t *hist;
  knnsum_t *s;
  unsigned int cur, low, high;

  hist = (knn_hist_t*)h;
  s = (knnsum_t*)elem;
  

  cur = hist->num_bins / 2;
  
  low = 0;
  high = hist->num_bins - 1;
  
  while (low < high){
    if (hist->degs[cur] < s->k){
      low = cur+1;
    }
    else if(hist->degs[cur] >= s->k){
      high = cur;
    }
    cur = (high + low) / 2;
  }

  hist->num[cur] += s->Nk * s->k;
  hist->values[cur] += s->knnsum;
}


/* print knn(k) for each value of k in the graph (i.e., without any
   binning) */
void print_knn_nobin(knntree_t t){
  
  iltree_view_pre(t);
}



/* print the values of knn(k) averaging over num_bins of equal size */
void print_knn_lin(knntree_t t, unsigned int num_bins){
  
  unsigned int kmin, kmax, step, i;
  knn_hist_t hist;
  
  if (!t){
    return;
  }


  kmin = ((knnsum_t*)iltree_getmin(t))->k;
  kmax = ((knnsum_t*)iltree_getmax(t))->k;
  
  hist.num_bins = num_bins;
  
  step = (kmax - kmin)/num_bins + 1;
  
  hist.degs = malloc(num_bins * sizeof(unsigned int));
  hist.num = malloc(num_bins * sizeof(unsigned int));
  hist.values = malloc(num_bins * sizeof(double));


  hist.degs[0] = kmin + step;
  hist.values[0] = 0;
  hist.num[0] = 0;

  for(i=1; i<num_bins; i++){
    hist.degs[i] = hist.degs[i-1] + step;
    hist.values[i] = 0;
    hist.num[i] = 0;
  }
  
  iltree_map_args(t, aggregate_knn, &hist);

  for(i=0; i<hist.num_bins; i++){
    if (hist.num[i])
      printf("%d %2.8g\n", hist.degs[i], hist.values[i] / hist.num[i] );
  }

  free(hist.degs);
  free(hist.num);
  free(hist.values);
  
}


/* print knn(k) using exponential (a.k.a. logarithmic) binning */

void print_knn_exp(knntree_t t, double alpha){
  
  unsigned int kmin, kmax, i;
  double width;
  knn_hist_t hist;
  
  if (!t){
    return;
  }

  
  kmin = ((knnsum_t*)iltree_getmin(t))->k;
  kmax = ((knnsum_t*)iltree_getmax(t))->k;
  
  
  hist.num_bins = (int)ceil(log(kmax)/log(alpha)) + 1;
  
  
  hist.degs = malloc(hist.num_bins * sizeof(unsigned int));
  hist.num = malloc(hist.num_bins * sizeof(unsigned int));
  hist.values = malloc(hist.num_bins * sizeof(double));

  width = 2;
  
  hist.degs[0] = kmin + width;
  hist.values[0] = 0;
  hist.num[0] = 0;

  for(i=1; i<hist.num_bins; i++){
    hist.values[i] = 0;
    hist.num[i] = 0;
    width *= alpha;
    hist.degs[i] = (int)ceil(hist.degs[i-1] + width);
  }
  
  iltree_map_args(t, aggregate_knn, &hist);

  for(i=0; i<hist.num_bins; i++){
    if(hist.num[i])
      printf("%d %2.8g\n", hist.degs[i], hist.values[i] / hist.num[i] );
  }
  
  free(hist.degs);
  free(hist.num);
  free(hist.values);
  
}



int main(int argc, char *argv[]){
  
  unsigned int N, K;
  unsigned int *J_slap=NULL, *r_slap=NULL;
  FILE *filein;
  char bin_type;
  unsigned int num_bins;
  double alpha;
  knntree_t t = NULL;
  
  

  if(argc < 2){
    usage(argv);
    exit(1);
  }
  
  if(!strcmp(argv[1], "-")){
    filein = stdin;
  }
  else{
    filein = openfile_or_exit(argv[1], "r", 2);
  }

  t = iltree_create(t);

  bin_type = NO_BIN;
  if (argc > 2){ /* the user has specified a binning type */
    if (!my_strcasecmp(argv[2], "lin")){
      bin_type = BIN_LIN;
      if(argc < 4){
        fprintf(stderr, "you must provide a number of bins for linear binning\n");
        exit(3);
      }
      num_bins = atoi(argv[3]);
    }
    else if (!my_strcasecmp(argv[2], "exp")){
      bin_type = BIN_EXP;
      if(argc < 4){
        fprintf(stderr, "you must provide an exponent for exponential binning\n");
        exit(4);
      }
      alpha = atof(argv[3]);
    }
  }
  
  read_slap(filein, &K, &N, &J_slap, &r_slap);
  fclose(filein);
  
  t = knntree_init(t, stdout);
  compute_knn(J_slap, r_slap, N, t);
  
  
  switch(bin_type){
  case BIN_LIN:
    print_knn_lin(t, num_bins);
    break;
  case BIN_EXP:
    print_knn_exp(t, alpha);
    break;
  default:
    print_knn_nobin(t);
    break;
  }
  iltree_destroy(t);
  free(J_slap);
  free(r_slap);
}
